datajunction-server/datajunction_server/construction/build.py

```diff
@@ -486,27 +457,55 @@ async def _build_tables_on_select(
                     dimensions,
                     access_control,
                 )
-                node_table = query_ast.select  # type: ignore
-                node_table.parenthesized = True  # type: ignore
                 memoized_queries[hash(node_query)] = query_ast
 
-        alias = amenable_name(node.name)
-        context = CompileContext(session=session, exception=DJException())
+            alias = amenable_name(node.name)
+            node_ast = ast.Alias(ast.Name(alias), child=query_ast.select, as_=True)  # type: ignore
+            query_ast.select.parenthesized = True  # type: ignore
+
+            filter_asts = build_filters(node, node_ast, filters)  # type: ignore
+        else:
+            alias = amenable_name(node.name)
+            node_ast = ast.Alias(ast.Name(alias), child=physical_table, as_=True)  # type: ignore
+            filter_asts = build_filters(node, physical_table, filters)
 
-        node_ast = ast.Alias(ast.Name(alias), child=node_table, as_=True)  # type: ignore
         for tbl in tbls:
-            if isinstance(node_ast.child, ast.Select) and isinstance(tbl, ast.Alias):
+            if isinstance(node_ast.child, ast.Select) and isinstance(tbl, ast.Alias):  # type: ignore
                 node_ast.child.projection = [
                     col
                     for col in node_ast.child.projection
                     if col in set(tbl.child.select.projection)
                 ]
+            if (
+                isinstance(node_ast.child, ast.Table)
+                and isinstance(tbl, ast.Aliasable)
+                and hasattr(tbl, "alias")
+                and tbl.alias
+            ):
+                node_ast.alias.name = tbl.alias
+
             await node_ast.compile(context)
+
+            # This will push down the filter expression to CTEs, subqueries, and top-level queries.
```

Comment:
@agorajek This block solves the issue with filter pushdowns not applying to the right `SELECT` clause. Basically, each filter expression applies to some `TableExpression`, and we need to append the filter expressions to the `TableExpression`'s nearest SELECT clause, not the overall SELECT clause. This makes it so that the table aliases are always applied correctly.